File: src/main/java/congard/agh/evolution/simulation/math/Rect.java
package congard.agh.evolution.simulation.math;

public class Rect {
    private final Point2d upperLeft;
    private final Point2d lowerRight;

    public Rect(Point2d upperLeft, Point2d lowerRight) {
        // TODO: check correctness
        this.upperLeft = upperLeft;
        this.lowerRight = lowerRight;
    }

    public int getWidth() {
        return lowerRight.getX() - upperLeft.getX();
    }

    public int getHeight() {
        return lowerRight.getY() - upperLeft.getY();
    }

    public Point2d getUpperLeft() {
        return upperLeft;
    }

    public Point2d getLowerRight() {
        return lowerRight;
    }

    public boolean contains(Point2d point2d) {
        var x1 = upperLeft.getX();
        var x2 = lowerRight.getX();
        var y1 = upperLeft.getY();
        var y2 = lowerRight.getY();
        var x = point2d.getX();
        var y = point2d.getY();

        return (x1 <= x && x <= x2) && (y1 <= y && y <= y2);
    }
}

File: src/main/java/congard/agh/evolution/gui/math/Rect.kt
package congard.agh.evolution.gui.math

class Rect(
    val upperLeft: Vec2,
    val lowerRight: Vec2
) {
    constructor(upperLeft: Vec2, width: Float, height: Float)
            : this(upperLeft, upperLeft + Vec2(width, height))

    constructor(upperLeft: Vec2, size: Float) : this(upperLeft, size, size)

    fun x() = upperLeft.x
    fun y() = upperLeft.y

    fun width() = lowerRight.x - upperLeft.x
    fun height() = lowerRight.y - upperLeft.y

    fun size() = lowerRight - upperLeft

    fun center() = upperLeft + size() / 2f

    operator fun contains(v: Vec2): Boolean {
        val x1 = upperLeft.x
        val x2 = lowerRight.x
        val y1 = upperLeft.y
        val y2 = lowerRight.y
        return v.x in x1..x2 && v.y in y1..y2
    }
}
